### ISSUE TITLE
emit JSON

### ISSUE THREAD

**Issue by cburgard** [biojppm/rapidyaml#20]
There is no JSON emitter as far as I can tell. Are there any plans to add one? It should be a minimal addition, as far as I can tell.

**Comment by biojppm** [comment]
I have no immediate plans to add one. I have other urgent things to look at, like the installer and conformance to the YAML test suite.

If you are in a rush, I would of course accept a PR implementing that functionality. The natural place to put that would be the [emit.hpp](https://github.com/biojppm/rapidyaml/blob/master/src/c4/yml/emit.hpp) header file. You can also take guidance from the code there.

**Comment by cburgard** [comment]
Thanks for the guidance. I had a look at the code and found that the only things that would really need to be changed would be the contents of `emit.def.hpp`, whereas the `emit.hpp` would mostly be copied.
I feel like the most elegant solution would be to template the contents of `emit.hpp` to some , write template specializations of the contents of `emit.hpp` for YML and JSON type output, and then instantiate the writers in `emit.hpp` to the corresponding types.
Also, I'm tempted to move the whole JSON stuff to a separate namespace, perhaps `c4/json` to avoid confusion.
What do you think about these plans? Would you prefer things to be done differently? If so, how?

**Comment by biojppm** [comment]
Something like what you say could work, with some adjustments. Let's focus first on making the code work, then we can refactor to tidy it up (regarding namespaces/headers/function names, etc).

There are three separate entities in that file: the writers, the emitter class and the emit() freestanding functions. 

 * The writers need no adjustment
 * For now, I would duplicate and work on the emitter class, which does most of the work, to adjust for emitting valid json. Let's call this EmitterJSON. Then at the top of the file we would have EmitterJSONBuf, etc. Since JSON is so much simpler than YAML, this class will be much shorter; the parts of YAML that are incompatible with JSON (eg, anchors and references) should be verified for on the visit functions. For now, let's make this an error if any such entity exists when emitting JSON.
* Finally, each freestanding function at the end of the file would have a emit_json() counterpart, placed next to it.

This will result in `c4::yml::emit_json()` which is indeed awkward, but we'll address that later.

The most important refactor will be in retaining the common parts of the emitter classes. But first, let's have it working; duplication there is no big deal since this is a transient situation, and when we have both yml and json emitters it will be obvious to see how to arrange the classes; maybe some form of CRTP would be best (and absolutely no virtual calls).

**Comment by cburgard** [comment]
Ok, I'll write down some preliminary version soon then.
A few more questions though:
- what should happen to the `std::ostream` operators? I don't suppose we can make JSON copies of those, but we do need some way of writing JSON to `std::ostream` (at least, that would be specifically my use case). What do you have in mind for that?
- I was hoping that we could somehow translate YAML references into JSON pointers, but that is something for later I suppose.

**Comment by biojppm** [comment]
> what should happen to the std::ostream operators? I don't suppose we can make JSON copies of those, but we do need some way of writing JSON to std::ostream

There are no changes required there. The writer `WriterOStream<std::ostream>` deals with that, and does not need to know anything about JSON or YAML. We only need to work at the Emitter level, which emits the formatted output to itstemplated writer.

**Comment by cburgard** [comment]
Yes, but the way it's coded now, the `<<` operator will always emit YAML, never JSON, so I was wondering whether there are any good options to make such an operator happen also for JSON.

**Comment by biojppm** [comment]
You're right! But I think there's an easy workaround which adheres to the spirit.

Here's the current version of the relevant part:
```c++
// always emits yaml
template<class OStream>
inline OStream& operator<< (OStream& s, Tree const& t)
{
    EmitterOStream<OStream> em(s);
    em.emit(t.rootref());
    return s;
}
```
I'm thinking that as a first approach we can use a tag type; something like this
```c++
struct as_json
{
    Tree const* tree;
    as_json(Tree const& t) : tree(&t) {}
};
template<class OStream>
inline OStream& operator<< (OStream& s, as_json const& t)
{
    EmitterJSONOStream<OStream> em(*s.tree); // instantiate a json emitter instead
    em.emit(t.rootref());
    return s;
}
```
which give or take could be used like so:
```c++
c4::yml::Tree tree;
std::cout << tree; // emits YAML
std::cout << as_json(tree); // emits JSON
```

**Comment by biojppm** [comment]
already merged in #22
